### PR TITLE
``_StrPath`` is dead; long live ``_StrPath``

### LINKED ISSUE #12686: QEMU fails to build with sphinx 8
### Describe the bug

https://bugs.gentoo.org/936620

```
[45/908] /usr/bin/env CONFDIR=/etc/qemu /usr/bin/sphinx-build -q -j auto -Dversion=9.0.2 -Drelease= -Ddepfile=docs/docs.d -Ddepfile_stamp=docs/docs.stamp -b html -d /var/tmp/portage/app-emulation/qemu-9.0.2/work/qemu-9.0.2/tools-build/docs/manual.p /var/tmp/portage/app-emulation/qemu-9.0.2/work/qemu-9.0.2/docs /var/tmp/portage/app-emulation/qemu-9.0.2/work/qemu-9.0.2/tools-build/docs/manual
FAILED: docs/docs.stamp 
/usr/bin/env CONFDIR=/etc/qemu /usr/bin/sphinx-build -q -j auto -Dversion=9.0.2 -Drelease= -Ddepfile=docs/docs.d -Ddepfile_stamp=docs/docs.stamp -b html -d /var/tmp/portage/app-emulation/qemu-9.0.2/work/qemu-9.0.2/tools-build/docs/manual.p /var/tmp/portage/app-emulation/qemu-9.0.2/work/qemu-9.0.2/docs /var/tmp/portage/app-emulation/qemu-9.0.2/work/qemu-9.0.2/tools-build/docs/manual

Extension error (depfile):
Handler <function write_depfile at 0x77a1775ff560> for event 'build-finished' threw an exception (exception: unsupported operand type(s) for +: 'PosixPath' and 'str')
```

### How to Reproduce

```python
"""depfile is a Sphinx extension that writes a dependency file for
   an external build system"""

import os
import sphinx
import sys
from pathlib import Path

__version__ = '1.0'

def get_infiles(env):
    for x in env.found_docs:
        yield env.doc2path(x)
        yield from ((os.path.join(env.srcdir, dep)
                    for dep in env.dependencies[x]))
    for mod in sys.modules.values():
        if hasattr(mod, '__file__'):
            if mod.__file__:
                yield mod.__file__
    # this is perhaps going to include unused files:
    for static_path in env.config.html_static_path + env.config.templates_path:
        for path in Path(static_path).rglob('*'):
            yield str(path)


def write_depfile(app, exception):
    if exception:
        return

    env = app.env
    if not env.config.depfile:
        return

    # Using a directory as the output file does not work great because
    # its timestamp does not necessarily change when the contents change.
    # So create a timestamp file.
    if env.config.depfile_stamp:
        with open(env.config.depfile_stamp, 'w') as f:
            pass

    with open(env.config.depfile, 'w') as f:
        print((env.config.depfile_stamp or app.outdir) + ": \\", file=f)
        print(*get_infiles(env), file=f)
        for x in get_infiles(env):
            print(x + ":", file=f)


def setup(app):
    app.add_config_value('depfile', None, 'env')
    app.add_config_value('depfile_stamp', None, 'env')
    app.connect('build-finished', write_depfile)

    return dict(
        version = __version__,
        parallel_read_safe = True,
        parallel_write_safe = True
    )
```

### Environment Information

```text
#######################################
# installed packages (qlist -ICvUSS): #
#######################################
dev-python/sphinx-8.0.0_rc1:0 -doc -latex python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinx-rtd-theme-2.0.0-r1:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-applehelp-1.0.8:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-devhelp-1.0.6:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-htmlhelp-2.0.6:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-jquery-4.1:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-jsmath-1.0.1-r3:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-qthelp-1.0.8:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
dev-python/sphinxcontrib-serializinghtml-1.1.10:0 python_targets_pypy3 python_targets_python3_10 python_targets_python3_11 python_targets_python3_12 -python_targets_python3_13 -test
```

### Sphinx extensions

_No response_

### Additional context

I'm pretty confident the root cause of this is https://github.com/sphinx-doc/sphinx/commit/de15d61a46daaaa2b0a0e341cdb4e0abe107e012

It changes app.env.doc2path to return a `pathlib.Path` instead of a `str`, breaking consumers that assume it's a str.

I also looked in the changelog and saw #12650 which indicates there are many other places that were formerly using `sphinx.util._pathlib._StrPath` as return values, which both provided a Path object, and allowed people to keep using it as a `str` with a deprecation warning. But doc2path never used _StrPath which means no warning was emitted when people used it as a str. Perhaps QEMU would have noticed, perhaps not.

### PR DESCRIPTION
**Pull request by AA-Turner**
Closes #12686.

### CODE REVIEW

File: sphinx/util/_pathlib.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+from pathlib import Path, PosixPath, PurePath, WindowsPath
+from typing import Any
+
+from sphinx.deprecation import RemovedInSphinx90Warning
+
+_STR_METHODS = frozenset(str.__dict__)
+_PATH_NAME = Path().__class__.__name__
+
+_MSG = (
+    'Sphinx 8 will drop support for representing paths as strings. '
```
Comment (smithdc1) [inline sphinx/util/_pathlib.py:28]: Should this say Sphinx 9?

Comment (AA-Turner) [reply sphinx/util/_pathlib.py:28]: Thanks, done in 46ce447a693e81f07a1320641b995d01d7987750.

A

**Comment by jfbu** [comment]
ping @AA-Turner with this commit I started seeing 

```
copying static files... /Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/filecmp.py:62: RemovedInSphinx90Warning: Sphinx 8 will drop support for representing paths as strings. Use "pathlib.Path" or "os.fspath" instead.
  outcome = _cache.get((f1, f2, s1, s2))
```

when building our own html docs.  I could not identify where the last line comes from, I thought a remnant of some debugging or may some doctest (but this commit did not touch doc/)

```text
Python version: 3.12.3 (CPython)
Docutils version: 0.20.1 
Pygments version: 2.18.0
Jinja2 version: 3.1.4
```

With 8.0.0rc1 I only see
```
copying static files... done
copying extra files... done
```

**Comment by AA-Turner** [comment]
#12708 hopefully fixes this.

A

**Pull request merged**